src/xunit.performance.api/Profilers/Etw/SimpleTraceEventParser.cs

```diff
@@ -0,0 +1,378 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Parsers;
+using Microsoft.Diagnostics.Tracing.Parsers.Clr;
+using Microsoft.Diagnostics.Tracing.Parsers.Kernel;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Xunit.Performance.Api.Profilers.Etw
+{
+    /// <summary>
+    /// Provides a simple interface to extract ETW process/modules data from an *.etl file.
+    /// </summary>
+    public sealed class SimpleTraceEventParser
+    {
+        /// <summary>
+        /// Gets profile data from the provided ScenarioInfo object
+        /// </summary>
+        /// <param name="scenarioInfo"></param>
+        /// <returns>A collection of the profiled processes for the given scenario.</returns>
+        /// <remarks>
+        /// Some assumptions:
+        ///     1. The scenario launches a single process, but itself can launch child processes.
+        ///     2. Process started/stopped within the ETW session.
+        /// </remarks>
+        public IReadOnlyCollection<Process> GetProfileData(ScenarioExecutionResult scenarioInfo)
+        {
+            var processes = new List<Process>();
+            Module defaultNtoskrnlModule = null;
+            var pmcSamplingIntervals = new Dictionary<int, long>();
+
+            Func<ProcessTraceData, bool> IsOurProcess = (ProcessTraceData obj) => {
+                return obj.ProcessID == scenarioInfo.ProcessExitInfo.ProcessId
+                    || processes.Any(process => process.Id == obj.ProcessID || process.Id == obj.ParentID);
+            };
+
+            using (var source = new ETWTraceEventSource(scenarioInfo.EventLogFileName))
+            {
+                if (source.EventsLost > 0)
+                    throw new Exception($"Events lost in trace '{scenarioInfo.EventLogFileName}'");
+
+                ////////////////////////////////////////////////////////////////
+                // Process data
+                var parser = new KernelTraceEventParser(source);
+                parser.ProcessStart += (ProcessTraceData obj) => {
+                    if (IsOurProcess(obj))
+                    {
+                        var process = new Process(obj.ImageFileName, obj.ProcessID, obj.ParentID);
+                        process.LifeSpan.Start = obj.TimeStamp;
+                        processes.Add(process);
+                    }
+                };
+                parser.ProcessStop += (ProcessTraceData obj) => {
+                    if (IsOurProcess(obj))
+                        processes.Single(process => process.Id == obj.ProcessID).LifeSpan.End = obj.TimeStamp;
```

Comment:
What will happen here if there are missed events and the process start was never processed?  I suspect it will throw a NRE in which case it might be good to check for this and provide a better error message.

Reply:
I throw an exception earlier if there are lost events. In addition, we only get here if the process has been added to the list.

---
In reply to: [146053767](https://github.com/Microsoft/xunit-performance/pull/242#discussion_r146053767) [](ancestors = 146053767)